### PR TITLE
feat: Compute matching patterns for automatic induction

### PR DESCRIPTION
**Pull request by RustanLeino**
This PR aims to help stabilize verification by filling in matching patterns for the quantifiers introduced by automatic induction to represent the induction hypothesis. It also suppresses the generation of the induction hypothesis if no such matching patterns are found.

### Full description

This PR computes matching patterns for the quantification that's about to be used with automatic induction. If there are no matching patterns, the induction hypothesis is not added. Tooltips or warnings show the patterns or announce the lack thereof.

The PR no longer uses arrow-typed variables as induction variables. (If a user really wants them, an `{:induction ...}` attribute can be given manually.)

Treat `this` more like other parameters when computing induction variables.

With this PR, ternary expressions (that is, `_ ==#[_] _` and `_ !=#[_] _`) are considered as candidate trigger expressions. In addition, a codatatype `==` (which is defined by Dafny as a greatest predicate) is considered as a focal predicate for extreme predicates.

The PR also fixes a crash in trigger selection when the candidate expression has a lambda expression.

Finally, the "selected trigger" tooltip is extended to also show the `t := e` binding for any bound variable added as part of a quantifier rewrite.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Rewriters/InductionRewriter.cs [hunk 1]
```diff
@@ -253,7 +289,10 @@ List<List<Expression>> ComputeInductionTriggers(List<Expression> inductionVariab
     var triggersCollector = new Triggers.TriggersCollector(finder.exprsInOldContext, Reporter.Options, moduleDefinition);
     var quantifierCollection = new Triggers.ComprehensionTriggerGenerator(quantifier, Enumerable.Repeat(quantifier, 1), Reporter);
     quantifierCollection.ComputeTriggers(triggersCollector);
-    var triggers = quantifierCollection.GetTriggers();
+    // Get the computed triggers, but only ask for those that do not require additional bound variables. (An alternative to this
+    // design would be to add {:matchinglooprewrite false} to "quantifier" above. However, that would cause certain matching loops
+    // to be ignored, so it is safer to not include triggers that require additional bound variables.)
+    var triggers = quantifierCollection.GetTriggers(false);
```
Comment (cpitclaudel) [inline Source/DafnyCore/Rewriters/InductionRewriter.cs:295]: From the comment above I don't immediately see why we don't need triggers for those additional bound variables.

**Review by cpitclaudel — COMMENTED** [review]
Here is a first pass with mostly cosmetic details.  I'm currently making my way through the core computation.

File: Source/DafnyCore/Rewriters/InductionRewriter.cs [hunk 2]
```diff
@@ -156,52 +165,149 @@ void ComputeInductionVariables<VarType>(IToken tok, List<VarType> boundVars, Lis
...
+      if (!(n.Type.IsTypeParameter || n.Type.IsAbstractType || n.Type.IsInternalTypeSynonym || n.Type.IsArrowType) &&
+          (args != null || InductionHeuristic.VarOccursInArgumentToRecursiveFunction(Reporter.Options, body, n))) {
         inductionVariables.Add(new IdentifierExpr(n.Tok, n));
       }
     }
 
     if (inductionVariables.Count != 0) {
+      List<List<Expression>> triggers = null;
+      if (lemma != null) {
+        triggers = ComputeInductionTriggers(inductionVariables, body, lemma.EnclosingClass.EnclosingModuleDefinition);
```
Comment (cpitclaudel) [inline Source/DafnyCore/Rewriters/InductionRewriter.cs:206]: I find the logic here a bit hard to follow: we have ComputeInductionTriggers and ComputeAndReportInductionTriggers and ReportInductionTriggers, and two sets of calls for each of these methods.  Is the duplication necessary?  Can we add a comment why?

File: Source/DafnyCore/Triggers/ComprehensionTriggerGenerator.cs [hunk 3]
```diff
@@ -230,5 +230,22 @@ internal void CommitTriggers(SystemModuleManager systemModuleManager) {
...
+      var triggers = new List<List<Expression>>();
+      foreach (var triggerWriter in partWriters) {
+        if (includeTriggersThatRequireNamedExpressions || triggerWriter.NamedExpressions.Count == 0) {
+          foreach (var triggerTerms in triggerWriter.Candidates) {
+            var trigger = new List<Expression>();
+            foreach (var triggerTerm in triggerTerms.Terms) {
+              trigger.Add(triggerTerm.Expr);
+            }
+
+            triggers.Add(trigger);
```
Comment (cpitclaudel) [inline Source/DafnyCore/Triggers/ComprehensionTriggerGenerator.cs:244]: (Reattaching this comment to the latest commit, I think this would read better as a Select, possibly wrapped in a SelectMany

File: Source/DafnyCore/Rewriters/InductionRewriter.cs [hunk 4]
```diff
@@ -156,52 +165,149 @@ void ComputeInductionVariables<VarType>(IToken tok, List<VarType> boundVars, Lis
...
+      // Next, look for matching patterns for the induction hypothesis.
+      if (lemma != null) {
+        var triggers = ComputeAndReportInductionTriggers(lemma, ref attributes, goodArguments, body);
+        if (triggers.Count == 0) {
+          var suppressWarnings = Attributes.Contains(attributes, "nowarn");
+          var warningLevel = suppressWarnings ? ErrorLevel.Info : ErrorLevel.Warning;
+
+          Reporter.Message(MessageSource.Rewriter, warningLevel, null, tok,
+            $"Could not find a trigger for the induction hypothesis. Without a trigger, this may cause brittle verification. " +
+            $"Change or remove the {{:induction}} attribute to generate a different induction hypothesis, or add {{:nowarn}} to silence this warning. " +
```
Comment (cpitclaudel) [inline Source/DafnyCore/Rewriters/InductionRewriter.cs:178]: Is there a way to make it produce a quantifier without triggers (and thus recover the previous behavior?)

**Review by cpitclaudel — CHANGES_REQUESTED** [review]
OK, I finished reading through.  It looks good to me!  Beyond stylistic concerns that I left in individual comments, my main worry is the impact that this will have on consumers: how many proofs will be broken, and how good a transition path are we giving them?  Is there an easy way to recover the previous behavior (e.g. what if we made :induction true do that, though with a warning?

File: Source/DafnyStandardLibraries/src/Std/Arithmetic/LittleEndianNat.dfy [hunk 5]
```diff
@@ -122,8 +122,17 @@ abstract module {:disableNonlinearArithmetic} Std.Arithmetic.LittleEndianNat {
...
+    assert DropFirst(xs1) == [];
+    calc {
+      ToNatRight(xs);
+      { reveal ToNatRight(); }
+      ToNatRight(xs1) * BASE() + First(xs);
+      { reveal ToNatRight(); }
+      (ToNatRight([]) * BASE() + First(xs1)) * BASE() + First(xs);
+      { reveal ToNatRight(); }
+      (0 * BASE() + First(xs1)) * BASE() + First(xs);
+    }
```
[the remark this answers is not in the document]
Comment (RustanLeino) [reply Source/DafnyStandardLibraries/src/Std/Arithmetic/LittleEndianNat.dfy:135]: I don't see how that rewrite helps any. It's nice to have a name for the expression `[xs[1]]`. But I removed the assertion `assert DropFirst(xs1) == [];`, which wasn't needed.

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/dafny3/Abstemious.dfy [hunk 6]
```diff
@@ -206,3 +206,25 @@ lemma Fib_Correct(n: nat)
...
+  ensures OhOnes().tail ==#[k] ones()
+{
+  // automatic induction on k
+}
+
+lemma Quantifier()
+  // the following quantifiers use the entire body as a trigger (previously, ternary expressions
+  // had not been considered as trigger candidates)
+  requires forall k: nat :: OhOnes().tail ==#[k] ones()
+  requires forall k: ORDINAL :: OhOnes().tail ==#[k] ones()
```
[the remark this answers is not in the document]
Comment (RustanLeino) [reply Source/IntegrationTests/TestFiles/LitTests/LitTest/dafny3/Abstemious.dfy:228]: You're right, it doesn't. But there is only one occurrence of `k` in each quantifier body, so I feel okay with the test as is.

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/triggers/let-expressions.dfy.expect [hunk 7]
```diff
@@ -1,6 +1,6 @@
 let-expressions.dfy(6,8): Info: Selected triggers: {s[i]}
 let-expressions.dfy(7,8): Info: Selected triggers: {s[i]}
-let-expressions.dfy(8,8): Info: Selected triggers: {s[_t#0], s[i]}
-let-expressions.dfy(9,8): Info: Selected triggers: {s[_t#0], s[i]}
+let-expressions.dfy(8,8): Info: Selected triggers: {s[_t#0], s[i]} where _t#0 := i + 1
+let-expressions.dfy(9,8): Info: Selected triggers: {s[_t#0], s[i]} where _t#0 := i + 1
```
[the remark this answers is not in the document]
Comment (RustanLeino) [reply Source/IntegrationTests/TestFiles/LitTests/LitTest/triggers/let-expressions.dfy.expect:4]: It seems I've been confused about the role of these names, and indeed about their worth as well. For now, I just back out my changes that had added printing of `where t#0 := ...` tooltips. Separately from this PR, we should rethink whether or not we really want to do these rewrites.

File: Source/DafnyCore/Verifier/Statements/BoogieGenerator.TrForallStmt.cs [hunk 8]
```diff
@@ -176,29 +175,35 @@ void TrForallStmtCall(IToken tok, List<BoundVar> boundVars, List<BoundedPool> bo
...
+          var arg = Substitute(s0.Args[i], null, substMap, s0.MethodSelect.TypeArgumentSubstitutionsWithParents());  // substitute the renamed bound variables for the declared ones
+          argsSubstMap.Add(s0.Method.Ins[i], new BoogieWrapper(initEtran.TrExpr(arg), s0.Args[i].Type));
+        }
         foreach (var ens in s0.Method.Ens) {
-          var p = Substitute(ens.E, receiver, argsSubstMap, s0.MethodSelect.TypeArgumentSubstitutionsWithParents());  // substitute the call's actuals for the method's formals
+          p = Substitute(ens.E, receiver, argsSubstMap, s0.MethodSelect.TypeArgumentSubstitutionsWithParents());  // substitute the call's actuals for the method's formals
           post = BplAnd(post, callEtran.TrExpr(p));
         }
-        tr = antitriggerBoundVarTypes;
+        tr = null;
```
[the remark this answers is not in the document]
Comment (RustanLeino) [reply Source/DafnyCore/Verifier/Statements/BoogieGenerator.TrForallStmt.cs:210]: I don't know. (And I don't know what this variable had once done. I had assumed the original purpose of the variable had gone away, but the variable was accidentally left in the code.)

File: Source/DafnyCore/Rewriters/InductionRewriter.cs [hunk 9]
```diff
@@ -195,6 +213,49 @@ void ComputeInductionVariables<TVarType>(IToken tok, List<TVarType> boundVars, E
...
 
+  /// <summary>
+  /// Obtain and return matching patterns for
+  ///     (forall inductionVariables :: body)
+  /// If there aren't any, then return null.
+  /// </summary>
+  List<List<Expression>> ComputeInductionTriggers(List<Expression> inductionVariables, Expression body, ModuleDefinition moduleDefinition) {
+    Contract.Requires(inductionVariables.Count != 0);
+
+    // Construct a quantifier, because that's what the trigger-generating machinery expects.
```
[the remark this answers is not in the document]
Comment (RustanLeino) [reply Source/DafnyCore/Rewriters/InductionRewriter.cs:262]: That would be better. When I started this PR, the resolver only computed the list of variables to use for the quantifier in the induction. The rest is done in the verifier (Boogie generator). I think one reason for that was that only the verifier knew how to construct the "decreases" antecedent. We could change this, since the source language now supports `decreases to` expression. But I suggest we would do such a refactoring in a different PR.

File: Source/DafnyCore/Triggers/ComprehensionTriggerGenerator.cs [hunk 10]
```diff
@@ -230,5 +230,19 @@ internal void CommitTriggers(SystemModuleManager systemModuleManager) {
...
+      var triggers = new List<List<Expression>>();
+      foreach (var triggerWriter in partWriters) {
+        foreach (var triggerTerms in triggerWriter.Candidates) {
+          var trigger = new List<Expression>();
+          foreach (var triggerTerm in triggerTerms.Terms) {
+            trigger.Add(triggerTerm.Expr);
+          }
+          triggers.Add(trigger);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (RustanLeino) [reply Source/DafnyCore/Triggers/ComprehensionTriggerGenerator.cs]: No, `triggers` is not flattened. It is a `List<List<Expression>>`, where the inner `List<Expression>` is the list of terms in one trigger.

I renamed `triggerTerms` `triggerCandidate` to improve the code a wee bit.

File: Source/DafnyCore/Rewriters/InductionRewriter.cs (hunk 1 above)
Comment (RustanLeino) [reply Source/DafnyCore/Rewriters/InductionRewriter.cs:295]: The idea was that if there are any `_t#0` rewrites, then we don't want to do automatic induction.

File: Source/DafnyCore/Rewriters/InductionRewriter.cs (hunk 2 above)
Comment (RustanLeino) [reply Source/DafnyCore/Rewriters/InductionRewriter.cs:206]: I added a comment. (The reason is to get the tooltips to come out in a nice order.)

File: Source/DafnyCore/Triggers/ComprehensionTriggerGenerator.cs (hunk 3 above)
Comment (RustanLeino) [reply Source/DafnyCore/Triggers/ComprehensionTriggerGenerator.cs:244]: I changed the inner `foreach` to use `Select` (actually, I changed it to use `ConvertAll`, which is like `Select` but returns a list instead of an enumeration).

**Comment by cpitclaudel** [comment]
Using `{:induction}` for backwards compatibility is tempting but I'm not sure it works: we have many cases of `{:induction}` already appearing in sources (mostly in quantifiers) in quantifiers.

We could use `{:induction "auto"}` to use the heuristic, plus `{:nowarn}`.  Alternatively, we could use `{:induction_triggers false}` to indicate that no triggers should be generated.

If we use `{:induction_triggers false}`, then it's natural to ask whether `{:induction_triggers}` should accept actual triggers as arguments.  I think it shouldn't: one should instead use an explicit `forall` expression or statement with regular triggers.  If we then add an editor action to generate this forall, then this will become even easier.

**Comment by cpitclaudel** [comment]
I've made a pass through this: 

- I renamed `:inductionPattern` to `:inductionTrigger`
- Passing `{:inductionTrigger}` (no trigger terms) restores the legacy behavior

Here are some examples:

```
predicate f(n: nat) { if n == 0 then true else f(n-1) }
predicate g(n: nat) { false }

// Default: auto-generated trigger. Proof works.
lemma Default(n: nat) ensures f(n) {}

// Manual list of variables. Proof works.
lemma {:induction n} ListOfVars(n: nat) ensures f(n) {}

// No induction. Proof fails.
lemma {:induction false} NoInduction(n: nat) ensures f(n) {}

// No induction, with manual proof. Proof works.
lemma {:induction false} ManualInduction(n: nat)
  ensures f(n)
{
  forall ih_n: nat | (n decreases to ih_n) {
    ManualInduction(ih_n);
  }
}

// No triggers, so no auto induction ⇒ Proof fails
lemma NoTriggers(n: nat) ensures f(n + 0) {}

// No triggers but forced induction, so warning. Proof works.
lemma {:induction} InductionWarning(n: nat) ensures f(n + 0) {}

// Explicit triggers, so no warning.  Proof works.
lemma {:induction} {:inductionTrigger f(n)} NoWarning2(n: nat) ensures f(n + 0) {}

// Legacy mode: auto induction with no triggers.  Proof works.
lemma {:inductionTrigger} Legacy(n: nat) ensures f(n) {}
lemma {:inductionTrigger} Legacy1(n: nat) ensures f(n + 0) {}
lemma {:induction} {:inductionTrigger} Legacy2(n: nat) ensures f(n + 0) {}
```

I updated the spec accordingly.  I still need to fix a printing issue.

**Comment by cpitclaudel** [comment]
I think this is ready for review.  We can do the `forall` statement generation in a separate PR.  I tried to split things into relevant commits, so reviewing them one by one might be best.

Also: I'm not sure how to regenerate the `.doo` files: running the recommended `make` command creates errors:

```
$ make -C Source/DafnyStandardLibraries update-binary
make: Entering directory 'Source/DafnyStandardLibraries'
dotnet run --project ../Dafny --no-build --roll-forward LatestMajor -- build -t:lib --hidden-no-verify=false  src/Std/dfyconfig.toml --output:build/DafnyStandardLibraries.doo
[…]
src/Std/Arithmetic/LittleEndianNat.dfy[ParametricConversion](113,30): Error: member 'First' does not exist in top-level module declaration '_default'
    |
113 |     ensures ToNatRight(xs) == First(xs)
    |                               ^^^^^
```

**Comment by RustanLeino** [comment]
@cpitclaudel Thanks for your help. I think everything has been addressed now.

**Review by cpitclaudel — APPROVED** [review]
The follow-up changes look good, thanks.  I merged with main.

**Review by cpitclaudel — APPROVED** [review]
